File: PolyScript/frameworks/test/test_compiler.zig
// Test Zig Compiler Tool for PolyScript Framework
// CRUD × Modes Architecture: Zero-boilerplate CLI development
// 
// Author: Mathew Burkitt, Dimension Technologies <mathew.burkitt@ditech.ai>

const std = @import("std");
const json = std.json;
const polyscript = @import("polyscript");

const TestCompilerTool = struct {
    pub const description = "Test Zig compiler tool demonstrating CRUD × Modes";
    
    pub fn create(resource: ?[]const u8, options: anytype, ctx: *polyscript.PolyScriptContext) !json.Value {
        try ctx.log("Compiling source...", "info");
        
        var obj = std.json.ObjectMap.init(ctx.allocator);
        try obj.put("compiled", json.Value{ .String = resource orelse "main.zig" });
        try obj.put("output", json.Value{ .String = if (resource) |r| 
            try std.mem.replaceOwned(u8, ctx.allocator, r, ".zig", "") 
        else "main" });
        try obj.put("optimized", json.Value{ .Bool = true });
        
        var buf: [64]u8 = undefined;
        const timestamp = try std.fmt.bufPrint(&buf, "{d}", .{std.time.timestamp()});
        try obj.put("timestamp", json.Value{ .String = timestamp });
        
        return json.Value{ .Object = obj };
    }
    
    pub fn read(resource: ?[]const u8, options: anytype, ctx: *polyscript.PolyScriptContext) !json.Value {
        _ = resource;
        _ = options;
        try ctx.log("Checking compilation status...", "info");
        
        var source_files = std.json.Array.init(ctx.allocator);
        try source_files.append(json.Value{ .String = "src/main.zig" });
        try source_files.append(json.Value{ .String = "src/utils.zig" });
        try source_files.append(json.Value{ .String = "build.zig" });
        
        var compiled_files = std.json.Array.init(ctx.allocator);
        try compiled_files.append(json.Value{ .String = "zig-out/bin/main" });
        try compiled_files.append(json.Value{ .String = "zig-cache/utils.o" });
        
        var missing = std.json.Array.init(ctx.allocator);
        try missing.append(json.Value{ .String = "zig-cache/build.o" });
        
        var obj = std.json.ObjectMap.init(ctx.allocator);
        try obj.put("source_files", json.Value{ .Array = source_files });
        try obj.put("compiled_files", json.Value{ .Array = compiled_files });
        try obj.put("missing", json.Value{ .Array = missing });
        
        var buf: [64]u8 = undefined;
        const timestamp = try std.fmt.bufPrint(&buf, "{d}", .{std.time.timestamp() - 7200});
        try obj.put("last_build", json.Value{ .String = timestamp });
        
        return json.Value{ .Object = obj };
    }
    
    pub fn update(resource: ?[]const u8, options: anytype, ctx: *polyscript.PolyScriptContext) !json.Value {
        _ = options;
        try ctx.log("Recompiling source...", "info");
        
        var obj = std.json.ObjectMap.init(ctx.allocator);
        try obj.put("recompiled", json.Value{ .String = resource orelse "main.zig" });
        try obj.put("reason", json.Value{ .String = "source file changed" });
        try obj.put("incremental", json.Value{ .Bool = true });
        
        var buf: [64]u8 = undefined;
        const timestamp = try std.fmt.bufPrint(&buf, "{d}", .{std.time.timestamp()});
        try obj.put("timestamp", json.Value{ .String = timestamp });
        
        return json.Value{ .Object = obj };
    }
    
    pub fn delete(resource: ?[]const u8, options: anytype, ctx: *polyscript.PolyScriptContext) !json.Value {
        _ = options;
        const res_name = resource orelse "build artifacts";
        try ctx.log(try std.fmt.allocPrint(ctx.allocator, "Cleaning {s}...", .{res_name}), "info");
        
        var cleaned = std.json.Array.init(ctx.allocator);
        try cleaned.append(json.Value{ .String = "zig-out/" });
        try cleaned.append(json.Value{ .String = "zig-cache/" });
        try cleaned.append(json.Value{ .String = ".zig-cache/" });
        
        var obj = std.json.ObjectMap.init(ctx.allocator);
        try obj.put("cleaned", json.Value{ .Array = cleaned });
        try obj.put("freed_space", json.Value{ .String = "45.8 MB" });
        
        var buf: [64]u8 = undefined;
        const timestamp = try std.fmt.bufPrint(&buf, "{d}", .{std.time.timestamp()});
        try obj.put("timestamp", json.Value{ .String = timestamp });
        
        return json.Value{ .Object = obj };
    }
};

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();
    
    try polyscript.PolyScriptTool(TestCompilerTool).run(allocator);
}

File: PolyScript/frameworks/test/test_compiler.jl
#!/usr/bin/env julia

#=
Test Julia Compiler Tool for PolyScript Framework
CRUD × Modes Architecture: Zero-boilerplate CLI development

Author: Mathew Burkitt, Dimension Technologies <mathew.burkitt@ditech.ai>
=#

# Add parent directory to load path
push!(LOAD_PATH, abspath(joinpath(@__DIR__, "../julia")))

# Include the framework directly
include("../julia/polyscript_framework.jl")
using .PolyScriptFramework
using JSON
using Dates

# Define the test compiler tool
struct TestCompilerTool <: PolyScriptTool end

# Tool description
PolyScriptFramework.description(::TestCompilerTool) = 
    "Test Julia compiler tool demonstrating CRUD × Modes"

# Create operation - compile source files
function PolyScriptFramework.create(::TestCompilerTool, resource, options, ctx)
    PolyScriptFramework.log_message(ctx, "Compiling $(something(resource, "source"))...", "info")
    
    resource_name = something(resource, "main.jl")
    output_name = replace(resource_name, ".jl" => "")
    
    return Dict(
        "compiled" => resource_name,
        "output" => output_name,
        "optimized" => get(options, "optimize", true),
        "timestamp" => Dates.format(now(), "yyyy-mm-dd HH:MM:SS")
    )
end

# Read operation - check compilation status
function PolyScriptFramework.read(::TestCompilerTool, resource, options, ctx)
    PolyScriptFramework.log_message(ctx, "Checking compilation status...", "info")
    
    return Dict(
        "source_files" => ["src/main.jl", "src/utils.jl", "src/config.jl"],
        "compiled_files" => ["main.so", "utils.so"],
        "missing" => ["config.so"],
        "last_build" => Dates.format(now() - Hour(2), "yyyy-mm-dd HH:MM:SS")
    )
end

# Update operation - recompile changed files
function PolyScriptFramework.update(::TestCompilerTool, resource, options, ctx)
    resource_name = something(resource, "source")
    PolyScriptFramework.log_message(ctx, "Recompiling $resource_name...", "info")
    
    return Dict(
        "recompiled" => something(resource, "main.jl"),
        "reason" => "source file changed",
        "incremental" => get(options, "incremental", true),
        "timestamp" => Dates.format(now(), "yyyy-mm-dd HH:MM:SS")
    )
end

# Delete operation - clean build artifacts
function PolyScriptFramework.delete(::TestCompilerTool, resource, options, ctx)
    resource_name = something(resource, "build artifacts")
    PolyScriptFramework.log_message(ctx, "Cleaning $resource_name...", "info")
    
    return Dict(
        "cleaned" => ["*.so", "*.ji", "compiled/", "pkgimages/"],
        "freed_space" => "22.3 MB",
        "timestamp" => Dates.format(now(), "yyyy-mm-dd HH:MM:SS")
    )
end

# Main entry point
if abspath(PROGRAM_FILE) == @__FILE__
    tool = TestCompilerTool()
    exit(PolyScriptFramework.run_tool(tool))
end